Add a Debug Controller (#2358)
kuzzleio/kuzzle: doc/2/guides/advanced/debugging/index.md

It's easier and safer to use the [Debug Controller](/core/2/api/controllers/debug) to debug Kuzzle remotely than exposing the debugger port using `node --inspect` since the [Debug Controller](/core/2/api/controllers/debug) is limited by the authentication mechanism of Kuzzle making it impossible to use if you don't have the proper rights to use it.
Whereas when launching Kuzzle with `node --inspect` you need to setup some Port Forwarding from your host machine to your local machine to be able to debug Kuzzle remotely, and you need to be sure that no one can access your instance using the Debug Port exposed.

:::warn
By default access to the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) through the [Debug Controller](/core/2/api/controllers/debug) is disabled, to use the methods of the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) you need to enable it in the [kuzzlerc file](https://github.com/kuzzleio/kuzzle/blob/master/.kuzzlerc.sample)

> by setting `security.debug.native_debug_protocol` to true and rebooting your instance.

Reviewer: ```suggestion
by setting `security.debug.native_debug_protocol` to `true` and rebooting your instance.
```

[Debug Modules](#debug-modules) are still accessible when `security.debug.native_debug_protocol` is disabled.
:::

## Debug Modules

Debug Modules are modules of Kuzzle that are exposed through the [Debug Controller](/core/2/api/controllers/debug) actions and allows you to better understand what is happening on your instance and helps you debug your instance live.

> Debug Modules have nothing to do with the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) and are perfectly safe to use since they do not expose the debugger directly, whereas the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) could be used to execute code remotely which is why access to the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) is disabled by default in the config and should only be enabled when you need to take memory snapshot, use breakpoint, the cpu profiler, etc.

Reviewer 2: ```suggestion
Debug Modules are not related to the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) and perfectly safe to use. They do not expose the debugger directly, whereas the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) could be used to execute code remotely which is why access to the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) is disabled by default. It should only be enabled when you need to take memory snapshot, use breakpoint, the cpu profiler, etc...
```

## Debug Kuzzle remotely using the Debug Controller and the Chrome Inspector

> ### Prerequisite

Reviewer: ```suggestion
### Prerequisites
```

To be able to debug Kuzzle remotely using the Chrome Inspector tool you need:

> - A user allowed to access to the every actions of the [Debug Controller](/core/2/api/controllers/debug), see the [Permissions](/core/2/main-concepts/permissions) page on how to setup the proper permissions.

Reviewer: ```suggestion
- A user allowed to access to all [Debug Controller](/core/2/api/controllers/debug)'s actions. See the [Permissions](/core/2/main-concepts/permissions) page on how to setup the proper permissions.
```

> - A user that has access to the every actions of the [Debug Controller](/core/2/api/controllers/debug), see the [Permissions](/core/2/main-concepts/permissions) page on how to setup the proper permissions.

Reviewer 2: ```suggestion
- A user allowed to access to the every actions of the [Debug Controller](/core/2/api/controllers/debug), see the [Permissions](/core/2/main-concepts/permissions) page on how to setup the proper permissions.
```

> - You need to set the config `security.debug.native_debug_protocol` to `true`, to allow the [Debug Controller](/core/2/api/controllers/debug) to execute actions from the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8), then restart your application. 

Reviewer: ```suggestion
- To set the config `security.debug.native_debug_protocol` to `true`, to allow the [Debug Controller](/core/2/api/controllers/debug) to execute actions from the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8), then restart your application. 
```

> - You need to have [Kourou](https://github.com/kuzzleio/kourou) the Kuzzle CLI installed. You can install it using `npm install -g kourou`.

Reviewer: ```suggestion
- [Kourou](https://github.com/kuzzleio/kourou), the Kuzzle CLI, installed. You can install it using `npm install -g kourou`.
```

> - You need to have the [Chrome Browser](https://www.google.com/intl/en_en/chrome/) installed.

Reviewer: ```suggestion
- The [Chrome Browser](https://www.google.com/intl/en_en/chrome/) installed.
```

### How to Debug

#### Step 1 - Setup proxy server

> To debug Kuzzle remotely using the Chrome Inspector you need to open a proxy server that will connects to Kuzzle and translate commands sent by the Chrome Inspector using under the format of the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) into actions of the [Debug Controller](/core/2/api/controllers/debug).

Reviewer 2: ```suggestion
To debug Kuzzle remotely using the Chrome Inspector, you'll need to open a proxy server connected to Kuzzle and translating commands sent by the Chrome Inspector using under the format of the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8) into actions of the [Debug Controller](/core/2/api/controllers/debug).
```

Reviewer: ```suggestion
To debug Kuzzle remotely using the Chrome Inspector, you need to open a proxy server that will connected to Kuzzle to translate commands sent by the Chrome Inspector with the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8)'s format into actions of the [Debug Controller](/core/2/api/controllers/debug).
```

> **Don't worry this is very easy to do**, you just have to use kourou to do that, just run

Reviewer 2: ```suggestion
**Don't worry this is very easy to do**, you just have to use Kourou to perform this:
```

```bash
kourou app:debug-proxy --host <Kuzzle Host> --port <Kuzzle Port> --username <username> --password <user password>
```

This will open a proxy server that will connects to your Kuzzle instance using the username and password given and will do the translation automatically.

:::warn
Remember the user should have the permissions to use all the actions of [Debug Controller](/core/2/api/controllers/debug) and you need to set the config `security.debug.native_debug_protocol` to `true`, to allow the [Debug Controller](/core/2/api/controllers/debug) to execute actions from the [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/v8).
:::

> Once the proxy server has started and has successfuly connect to your Kuzzle instance will see a message like this in your terminal:

Reviewer 2: ```suggestion
Once the proxy server has started and successfuly connects to your Kuzzle instance, you should see a message like this in your terminal:
```

Reviewer: ```suggestion
Once the proxy server has started and is successfully connected to your Kuzzle instance, you will will see a message like this in your terminal:
```

```bash
[ℹ] Connecting to ws://<host>:<port> ...
[ℹ] Connected to Kuzzle node: knode-lackadaisical-archimedes-79903
[✔] Listening on port 9222, forwarding to Kuzzle at <host>:<port>
[ℹ] Showing to Chrome Debugger as "Kuzzle Debugger - <host>:<port>"
[ℹ] Waiting for Chrome Debugger to connect...
```

```bash
[✔] Listening on port 9222, forwarding to Kuzzle at <host>:<port>
```

> You can see here that the proxy server is listening on port `9222` this will be needed for the next step.

Reviewer 2: ```suggestion
You can see here that the proxy server is listening on port `9222`. This will be needed for the next step.
```

#### Step 2 - Configure Chrome Inspector

> The proxy server is correctly installed, you now need to open the Chrome Inspector to debug your instance. To do so, open this url `chrome://inspect` in [Chrome Browser](https://www.google.com/intl/en_en/chrome/).

Reviewer: ```suggestion
The proxy server is correctly running, you now need to open the Chrome Inspector to debug your instance. To do so, open the url `chrome://inspect` in the [Chrome Browser](https://www.google.com/intl/en_en/chrome/).
```

> The proxy server is now setuped, you now need to open the Chrome Inspector to debug your instance, to do that you can open this url `chrome://inspect` in [Chrome Browser](https://www.google.com/intl/en_en/chrome/).

Reviewer 2: ```suggestion
The proxy server is correctly installed, you now need to open the Chrome Inspector to debug your instance. To do so, open this url `chrome://inspect` in [Chrome Browser](https://www.google.com/intl/en_en/chrome/).
```

> This will show you the Chrome debugger interface which will then be used to debug the Kuzzle instance.

Reviewer: ```suggestion
This will show you the Chrome debugger interface which will be used to debug the Kuzzle instance.
```

> This will open you the debug interface of Chrome which will then be used to debug the Kuzzle instance.

Reviewer 2: ```suggestion
This will show you the Chrome debugger interface which will then be used to debug the Kuzzle instance.
```

> To allow the Chrome Inspector to see our proxy server exposed by the command `kourou app:debug-proxy`

Reviewer: ```suggestion
To allow the Chrome Inspector to see our proxy server exposed by the command `kourou app:debug-proxy`,
```

> we need to add `localhost:9222` to the list of targets the Chrome Inspector can see and debug.

Reviewer: ```suggestion
you need to add `localhost:9222` to the list of targets the Chrome Inspector can see and debug.
```

:::info

> Here we are using the port `9222` but keep in mind that this port can be any other number and you need to change it accordingly on what port the proxy server is listening to, see end of [Step 1 - Setup proxy server](#step-1---setup-server-proxy)

Reviewer: ```suggestion
Here we are using the port `9222` but keep in mind that this port can be any other number. You need to change it accordingly on what port the proxy server is listening to, see end of [Step 1 - Setup proxy server](#step-1---setup-server-proxy)
```

The host will always be localhost since it's the IP of the proxy server that is running on your local machine.
:::

**1 -** Click on the `Configure` button, a dialog will appear where you can enter the IP and Port of your local proxy server.
![Image Configure Chrome Inspector Step 1](./config_step_1.png)

> **2 -** Enter your local target, here  `localhost:9222`

Reviewer: ```suggestion
**2 -** Enter your local target, here `localhost:9222`
```

**3 -** Confirm
![Image Configure Chrome Inspector Step 2](./config_step_2.png)

> **If everything worked so far you should see a new Remote Target called** `Kuzzle Debugger - <Instance Host>:<Instance Port>`.

Reviewer: ```suggestion
**If everything worked so far, you should see a new Remote Target called** `Kuzzle Debugger - <Instance Host>:<Instance Port>`.
```

![Image Configure Chrome Inspector Step 3](./config_step_3.png)

> **4 -** You can now click on `inspect` and your done, you can now debug your instance remotely using the Chrome Inspector.

Reviewer: ```suggestion
**4 -** You can now click on `inspect` and you're done.
```
Redundancy with next section

#### Step 3 - Debugging

You can now fully use the Chrome Inspector to debug you instance.

#### Step 4 - After Debugging

:::warn

> Don't forget to set the `security.debug.native_debug_protocol` configuration setting back to `false` once you're done debugging to increase the security of your instance in case of one of your admin users with access to the [Debug Controller](/core/2/api/controllers/debug) gets compromised.

Reviewer: ```suggestion
Don't forget to set the `security.debug.native_debug_protocol` configuration setting back to `false` once you're done debugging to increase the security of your instance. Doing so, if one of your admin users with access to the [Debug Controller](/core/2/api/controllers/debug) gets compromised, it won't harm your application.
```

> Don't forget to set the config `security.debug.native_debug_protocol` back to `false` once your done debugging to increase the security of your instance in case of one of your admin users with access to the [Debug Controller](/core/2/api/controllers/debug) gets compromised.

Reviewer 2: ```suggestion
Don't forget to set the `security.debug.native_debug_protocol` configuration setting back to `false` once you're done debugging to increase the security of your instance in case of one of your admin users with access to the [Debug Controller](/core/2/api/controllers/debug) gets compromised.
```
